feat: let the user cancel ongoing payments

rosvik opened the pull request:
_Might make sense to review one commit at a time_

This PR allows the user to cancel ongoing payments, and prevents several purchase flows going at once. 

Also sets a `isUser` query parameter for all cancel calls, indicating whether or not it was manually triggered by the user. This should allow us to monitor the amount of people who use the button.

https://github.com/user-attachments/assets/b850216c-88f6-4316-85fe-f4a4057dc629

closes https://github.com/AtB-AS/kundevendt/issues/20293

rosvik commented on src/ticketing/api.ts:
```diff
@@ -170,11 +170,12 @@ export async function reserveOffers({
 ): Promise<void> {
-  const url = `ticket/v3/payments/${payment_id}/transactions/${transaction_id}/cancel`;
-  await client.put(url, {}, {authWithIdToken: true});
+  const url = `ticket/v3/payments/${paymentId}/transactions/${transactionId}/cancel?isUser=${isUser}`;
+  await client.put(url, undefined, {authWithIdToken: true});
```
☝️ FYI @Nisha1306 and @adriansberg

adriansberg commented on src/stacks-hierarchy/Root_PurchaseConfirmationScreen/Root_PurchaseConfirmationScreen.tsx:
```diff
@@ -297,61 +295,101 @@ export const Root_PurchaseConfirmationScreen: React.FC<Props> = ({
+
+  return (
+    <Button
+      expanded={true}
```
```suggestion
      expanded
```

rosvik replied:
The app code mostly has `={true}` set for props like this. I'm open to changing that convention as bigger refactor (maybe an ESLint rule for one or the other), but for now I'll stick with this.

adriansberg replied:
Ah, TIL! Let's stick to that convention then.

adriansberg commented:
Based on the recording in the description, even though the user cancels payment the ticket is still processed and active?

marius-at-atb commented on src/stacks-hierarchy/Root_PurchaseConfirmationScreen/Root_PurchaseConfirmationScreen.tsx:
```diff
@@ -112,6 +120,12 @@ export const Root_PurchaseConfirmationScreen: React.FC<Props> = ({
   });
   const cancelPaymentMutation = useCancelPaymentMutation();
+  useEffect(() => {
+    if (cancelPaymentMutation.isSuccess) {
+      cancelPaymentMutation.reset();
+      reserveMutation.reset();
+    }
```
To avoid the useEffect, this function can be passed in to `onSuccess` in `useCancelPaymentMutation` instead. (Same as in e.g. `useSendSupportRequestMutation`)

rosvik replied:
Ah, thanks! That's much cleaner.

marius-at-atb commented:
@adriansberg Good question - is it possible to send a cancel callback or listener to the pop up? When does a user not want to cancel the payment, having already pressed cancel once?

rosvik commented:
> Based on the recording in the description, even though the user cancels payment the ticket is still processed and active?

The in app browser we use for NETS doesn't seem to support a `onClose` callback, which means I haven't found a way to cancel the payment automatically. Please let me know if I've missed something 😅 https://github.com/proyecto26/react-native-inappbrowser

This flow makes more sense for Vipps, where the app is waiting for payments in the background while Vipps is open. If you then return to the app before completing payment, it makes more sense that you have to manually cancel.

adriansberg commented:
> This flow makes more sense for Vipps, where the app is waiting for payments in the background while Vipps is open. If you then return to the app before completing payment, it makes more sense that you have to manually cancel.

That sounds like it makes a lot more sense, yes. Been experiencing that confusion myself when the Vipps payment isn't completed.

marius-at-atb commented:
https://github.com/proyecto26/react-native-inappbrowser/issues/415#issuecomment-1647860408
Haven't checked, but this might work for onClose handling?

The Vipps case makes sense, so I guess we want to include the solution from this PR anyway :+1:

rosvik commented:
> Haven't checked, but this might work for onClose handling?

Aaahh, I've been looking in the entirely wrong place 🤦 Thanks!

rosvik commented:
We try again! Added a little refactor in 28621b453b2d9ed47a18bdb4d8cfa8d0db6ed82d, and added a `onCancel` callback in 70b34d1f3ef2867d46016837cb92f2eeb60571a0 ✨ 

https://github.com/user-attachments/assets/35b04444-f600-4903-8e8c-9f6025b3c5e2

marius-at-atb commented on src/stacks-hierarchy/Root_PurchaseConfirmationScreen/Root_PurchaseConfirmationScreen.tsx:
```diff
@@ -297,64 +302,159 @@ export const Root_PurchaseConfirmationScreen: React.FC<Props> = ({
+  totalPrice: number;
+  mode: 'TravelSearch' | 'Ticket' | undefined;
+  onCloseFocusRef: RefObject<any>;
+  reserve: OfferReservation | undefined;
```
Could we find a name/solution that makes the difference between `OfferReservation` and `Reservation` more clear here?
Atm. `reserveMutation.data` is used both as `reservation` for `cancelPaymentMutation.mutate`, and `reserve` here, which I didn't intuitively understand 🤔

rosvik replied:
Good point! Looks like `ReserveOffer` is used in the new sales service, so we could start using that for naming. https://github.com/AtB-AS/sales/blob/596055de91dc22b05b5b0c243993b75b467cb0d2/sales-service/src/handlers/sales/reserve.rs#L85

marius-at-atb reviewed (approved):
Nice work!

rosvik commented:
@adriansberg I may have made you some more conflicts with this 🙈 7d99ea5d8183c36001577ab27595de053b585838
